signers/legacy/search.go

```diff
@@ -0,0 +1,238 @@
+	}
+
+	return &SearchSigner{
+		Digester: &signers.Digester{
+			Digest: digest,
+		},
+		Identifiable: &signers.Identifiable{
+			IdRegex: re,
+		},
+		respSigner: NewSearchResponseSigner(digest),
+	}, nil
+}
+
+// This is not used, required by the interface
+func (v *SearchSigner) HashBody(r *http.Request) (string, *signers.AuthenticationError) {
+	data, err := signers.ReadBody(r)
+	if err != nil {
+		return "", signers.Errorf(500, signers.ErrorTypeInternalError, "Failed to read request body: %s", err.Error())
+	}
+	return v.HashBytes(data), nil
+}
+
+// This is not used, required by the interface
+func (v *SearchSigner) HashBytes(b []byte) string {
+	//h := sha256.New()
+	//h.Write(b)
+	//return base64.StdEncoding.EncodeToString(h.Sum(nil))
+	return ""
+}
+
+func (v *SearchSigner) GetResponseSigner() signers.ResponseSigner {
+	return v.respSigner
+}
+
+func (v *SearchSigner) Sign(r *http.Request, authHeaders map[string]string, secret string) (string, *signers.AuthenticationError) {
+
+	var hash string
+	var path_and_query string
+	var request_time int64
+
+	// get / validate headers
+	auth_headers := v.ParseAuthHeaders(r)
+    if auth_headers["acquia_solr_time"] == "" {
+		return "", signers.Errorf(403, signers.ErrorTypeMissingRequiredHeader, "Missing required cookie: acquia_solr_time")
+    }
+    if auth_headers["acquia_solr_nonce"] == "" {
+		return "", signers.Errorf(403, signers.ErrorTypeMissingRequiredHeader, "Missing required cookie: acquia_solr_nonce")
+    }
+    if auth_headers["acquia_solr_hmac"] == "" {
+		return "", signers.Errorf(403, signers.ErrorTypeMissingRequiredHeader, "Missing required cookie: acquia_solr_hmac")
+    }
+
+    request_time = time.Now().Unix()
+
+	body, err := ioutil.ReadAll(r.Body)
+    if err != nil {
+		return "", signers.Errorf(500, signers.ErrorTypeInternalError, "Failed to read request body: %s", err.Error())
+    }
+
+    if r.Method == "POST" {
+	    hash = generateSignature(string(body), request_time, secret)
+	    logger.Print("body: " + string(body))
+    } else {
+        path_and_query = r.URL.Path + "?" + r.URL.RawQuery;
+        hash = generateSignature(path_and_query, request_time, secret);
+        logger.Print("Path Requested: " + path_and_query)
+    }
+
+    return hash, nil
+}
+
+
+func ParseAuthHeadersSearch(r *http.Request) map[string]string {
+	auth_headers := map[string]string{}
+	auth_fields := []string {
+		"acquia_solr_time",
+		"acquia_solr_nonce",
+		"acquia_solr_hmac",
+	}
+
+	for _, field_name := range auth_fields {
+	    auth_cookie, err := r.Cookie(field_name)
+	    if err != nil {
+			logger.Print("Error retrieving:", field_name)
+		} else {
+			auth_headers[field_name] = auth_cookie.Value
+			logger.Print(field_name, ": ", auth_cookie.Value)
+		}
+    }
+    return auth_headers
+}
+
+func (v *SearchSigner) ParseAuthHeaders(req *http.Request) map[string]string {
+	return ParseAuthHeadersSearch(req)
```

Comment:
This does look silly, but I think it's the best we can do without changing the interface, which I wanted to avoid for now.  The trick is that the interface requires ParseAuthHeaders, but then the response signing code also needs to use ParseAuthHeaders, so a simple function needs to be exported that can do the same work without the SearchSigner object.  We can't just export ParseAuthHeaders, because then that conflicts with other implementations, like Dice's did with search until I changed it.  There's almost certainly a better design that prevents all this, removing this method from the interface is one option.